pkg/counter/counter.go

```diff
@@ -1,96 +1,84 @@
 package counter
 
 import (
+	"errors"
 	"fmt"
+	"math"
 	"strconv"
 	"sync/atomic"
-
-	"github.com/aquasecurity/tracee/pkg/errfmt"
 )
 
 type Counter struct {
-	count uint64
+	value uint64
 }
 
-func NewCounter(v uint64) Counter {
-	return Counter{
-		count: v,
-	}
+// NewCounter creates a new counter with the given initial value.
+func NewCounter(initialValue uint64) Counter {
+	return Counter{value: initialValue}
 }
 
-func (c *Counter) Increment(amount ...uint64) error {
-	if len(amount) == 0 {
-		return c.incAtomic(1)
-	}
-
-	sum, err := sumUint64(amount...)
-	if err != nil {
-		return errfmt.WrapError(err)
-	}
-
-	return c.incAtomic(sum)
+// Increase increases the counter by the given value (thread-safe).
+func (c *Counter) Increase(values ...uint64) error {
+	_, err := c.IncreaseAndRead(values...)
+	return err
 }
 
-func (c *Counter) Decrement(amount ...uint64) error {
-	if len(amount) == 0 {
-		return c.decAtomic(1)
+// IncreaseAndRead increases the counter by the given value (thread-safe) and returns the new value.
+func (c *Counter) IncreaseAndRead(values ...uint64) (uint64, error) {
+	if len(values) == 0 {
+		values = append(values, 1)
 	}
 
-	sum, err := sumUint64(amount...)
-	if err != nil {
-		return errfmt.WrapError(err)
-	}
+	var n uint64
 
-	return c.decAtomic(sum)
-}
-
-func (c *Counter) Read() uint64 {
-	return atomic.LoadUint64(&c.count)
-}
+	for _, value := range values {
+		n = atomic.AddUint64(&c.value, value)
+		if n < value {
+			return n, errors.New("counter overflow")
+		}
+	}
 
-func (c *Counter) Set(val uint64) {
-	atomic.StoreUint64(&c.count, val)
+	return n, nil // return last known value (atomically incremented)
 }
 
-func (c Counter) Format(f fmt.State, r rune) {
-	f.Write([]byte(strconv.FormatUint(c.count, 10)))
+// Decrease decreases the counter by the given value (thread-safe).
+func (c *Counter) Decrease(values ...uint64) error {
+	_, err := c.DecreaseAndRead(values...)
+	return err
 }
 
-func (c *Counter) incAtomic(v uint64) error {
-	if (v + c.count) < c.count {
-		return errorCounterWrapAround()
+// DecreaseAndRead decreases the counter by the given value (thread-safe) and returns the new value.
+func (c *Counter) DecreaseAndRead(values ...uint64) (uint64, error) {
+	if len(values) == 0 {
+		values = append(values, 1)
 	}
-	atomic.AddUint64(&c.count, v)
 
-	return nil
-}
+	var n uint64
 
-func (c *Counter) decAtomic(v uint64) error {
-	v = ^uint64(v - 1)
-	if (v + c.count) > c.count {
-		return errorCounterWrapAround()
+	for _, value := range values {
+		if value == 0 {
+			continue
+		}
+		n = atomic.AddUint64(&c.value, ^uint64(value-1))
+		if n == math.MaxUint64 {
+			return n, errors.New("counter underflow")
+		}
```

Comment:
Decrementing is tricky indeed. As it is now it only considers a wrap of 1 unity. When value is larger than 1, and a wrap occurs, n can be any value greater than the previous value. So this check won't fail even a wrap occurring.

Checking like this won't be a solution either, since we're not locking the whole arithmetic.

```go
	for _, value := range values {
		if value == 0 {
			continue
		}

		originalValue = atomic.LoadUint64(&c.value)
		n = atomic.AddUint64(&c.value, ^uint64(value-1))

		// Detecting underflow by comparing the new value with the original value
		if n > originalValue {
			// Reset the counter to its original value before returning the error.
			atomic.StoreUint64(&c.value, originalValue)
			return n, errors.New("counter underflow")
		}
	}
```

I suspect that case for that we need a mutex.

Reply:
Continue with my previous comment - if we first sum the values and then decrement we can avoid having a mutex. I think the previous implementation was good.

Reply:
In this case I suspect we need to check the values sum against wrapping as well.

Reply:
You still have a race there (the same you describe below about context switching). You're reading c.count without an underlaying atomic instruction and the CPU cache coherence might not have updated all CPUs caches (thus the value will be wrong). There is no free lunch, without holding a mutex there can't be LOAD and STORES (multiple operations). That is why a single atomic operation needs to change the value and give previous value.